lib/services/role.go

```diff
@@ -1299,14 +1300,29 @@ func (set RoleSet) CheckAccessToServer(login string, s Server) error {
 		}
 		matchLogin, loginMessage := MatchLogin(role.GetLogins(Allow), login)
 		if matchNamespace && matchLabels && matchLogin {
-			return nil
+			if role.GetOptions().RequireSessionMFA && !mfaVerified {
+				log.WithFields(log.Fields{
+					trace.Component: teleport.ComponentRBAC,
+				}).Debugf("Access to node %q denied, role %q requires per-session MFA; match(namespace=%v, label=%v, login=%v)",
+					s.GetHostname(), role.GetName(), namespaceMessage, labelsMessage, loginMessage)
+				return trace.AccessDenied("access to server requires MFA")
+			}
+			// Check all remaining roles, even if we found a match.
```

Comment:
Meaning that the first requireSessionMFA and unverified MFA is an access denied even though there was a match?
It feels a bit clumsy since if MFA is verified - you don't need to go through the rest of the roles - maybe account for that:
```go
if mfaVerified {
	return nil
}
allowed = true
continue

```

Reply:
yeah, from https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md#per-role:
> Generally, if at least one role that grants access to a resource (SSH node, k8s cluster, etc.) sets require_session_mfa: true, then MFA check is required. It's required even if another role grants access to the same resource without MFA.

Changed to return early if `mfaVerified` is set.